File: backend/app/core/upgrades/stat_modifiers/explosion_damage_3.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class ExplosionDamage3(UpgradeImplementation):
    id = 'explosion_damage_3'
    name = 'Volatile Core'
    description = '+25 explosion damage'
    rarity = 'legendary'
    type = 'stat_modifier'
    cost = 40

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 25
        appliers.modifiers.add_multiplier('bullet', 'explosionDamage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/explosion_damage_3.ts
import { Upgrade, type UpgradeDef, type ExplosionSpec } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const ExplosionDamage3Def: UpgradeDef = {
  id: "explosion_damage_3",
  name: "Volatile Core",
  description: "+25 explosion damage",
  rarity: RarityID.Legendary,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 40,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.ExplosionDamage,
  value: 25,
  isMultiplier: false,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 99999,
  dependentOn: [{ ids: ["explosive_bullets", "explosion_on_kill"] }],
}

export class ExplosionDamage3 extends Upgrade {
  // Stat fields in the def are UI/backend metadata only — the behavior is the
  // hook below, so the default modifier-channel application is skipped.
  onApply(): void {}

  modifyExplosion(explosion: ExplosionSpec): void {
    explosion.damage += this.def.value!
  }
}
